pkg/cmd/server/api/helpers.go

```diff
@@ -317,6 +331,62 @@ func GetAPIClientCertCAPool(options MasterConfig) (*x509.CertPool, error) {
 	return cmdutil.CertPoolFromFile(options.ServingInfo.ClientCA)
 }
 
+// GetNamedCertificateMap returns a map of strings to *tls.Certificate, suitable for use in tls.Config#NamedCertificates
+// Returns an error if any of the certs cannot be loaded
+// Returns nil if len(namedCertificates) == 0
+func GetNamedCertificateMap(namedCertificates []NamedCertificate) (map[string]*tls.Certificate, error) {
+	if len(namedCertificates) == 0 {
+		return nil, nil
+	}
+	namedCerts := map[string]*tls.Certificate{}
+	for _, namedCertificate := range namedCertificates {
+		cert, err := tls.LoadX509KeyPair(namedCertificate.CertFile, namedCertificate.KeyFile)
+		if err != nil {
+			return nil, err
+		}
+		for _, name := range namedCertificate.Names {
+			if _, exists := namedCerts[name]; exists {
+				return nil, fmt.Errorf("multiple certificates specified the name %q", name)
+			}
+			namedCerts[name] = &cert
+		}
+	}
+	return namedCerts, nil
+}
+
+// GetCertificateFunc returns a function that can be used in tls.Config#GetCertificate
+// Returns an error if any of the certs cannot be loaded
+// Returns nil if len(certs) == 0
+func GetCertificateFunc(certs map[string]*tls.Certificate) func(*tls.ClientHelloInfo) (*tls.Certificate, error) {
+	if len(certs) == 0 {
+		return nil
+	}
+	return func(clientHello *tls.ClientHelloInfo) (*tls.Certificate, error) {
+		if clientHello == nil || len(clientHello.ServerName) == 0 {
+			return nil, nil
+		}
+
+		name := clientHello.ServerName
+		name = strings.ToLower(name)
+		name = strings.TrimRight(name, ".")
+
+		if cert, ok := certs[name]; ok {
+			return cert, nil
+		}
+
+		// try replacing labels in the name with wildcards until we get a match.
+		labels := strings.Split(name, ".")
+		for i := range labels {
+			labels[i] = "*"
+			candidate := strings.Join(labels, ".")
```

Comment:
This would result in looking for "foo.bar.example.com" with "*.*.example.com", but don't we want "*.example.com"?  I didn't think I had to specify multiple wildcards counting off the subdomains.